packages/jaeger-ui/src/model/ddg/DdgEVManager.tsx

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { TEdge } from '@jaegertracing/plexus/lib/types';
+
+import { compareKeys } from './visibility-key';
+
+import { EDdgEdgeKeys, PathElem, DdgVertex, TDdgModel } from './types';
+
+export default class DdgEVManager {
+  private edges: Set<TEdge>;
+  private edgeToFarSideOfEdgePathElems: Map<TEdge, Set<PathElem>>;
+  private farSideOfEdgePathElemsToEdge: Map<Set<PathElem>, TEdge>;
+  private prevVisKey: string;
+  private pathElemToFarSideOfEdgePathElems: Map<PathElem, Set<PathElem>>;
+  private pathElemToVertex: Map<PathElem, DdgVertex>;
+  private vertices: Map<string, DdgVertex>;
+  private visIdxToPathElem: PathElem[];
```

Comment:
Also, there is a lot of redundancy in the data. I think this adds to the complexity. It's an optimization that facilitates quick access but at the cost of maintaining more state, which needs to be aligned during updates, etc. As opposed to having a simpler representation of the state, which would make updates easier to implement and follow (albeit with more traversing instead of accessing via an index or key).

Reply:
yeah it's definitely heavy in the amount of maps and bi-directional mappings. it adds complexity, but I think it will be required for performance when changing the visibility of many `pathElem`s at once, such as when hiding children of `pathElem` that has had its descendants fully expanded.

Reply:
While falling asleep last night I realized that I can remove the interim step of `pathElemToFarSideOfEdgeElems: Map<pathElem, Set<PathElem>>`
edges can be associated with pathElems exactly like vertices are. `pathElemToEdge: Map<PathElem, TEdge>`, and back: `edgeToFarSideElems: Map<TEdge, Set<PathElem>>`

Furthermore, I realized that `this.edges` can be supplanted by `this.edgeToFarSideElems.keys()`

Reply:
currently, these `private` properties are simply alphabetical. I think with the pruning in the previous comment and some grouping with comments it becomes fairly manageable:

```
// PathElems given to constructor that serve as underlying data of a graph
private readyonly visIdxToPathElem: PathElem[];

// over the lifespan of a manager, it is provided changing keys/codecs. The last processed key/codec is stored in order to avoid creating every collection of edges & vertices from scratch when large parts may be reused / unchanged.
private prevVisKey: string;

// Non-focal pathElems need to be connected to the focalNode. If non-focal pathElems are hidden, we need to know which edges may be removed.
private pathElemToEdge: Map<PathElem, TEdge>;
// Since a single edge may be shared by multiple pathElems, we need to track them all so that hiding one pathElem doesn't result in an edge being removed if other pathElems for that same edge are still visible
private edgeToFarSideElems: Map<TEdge, Set<PathElem>>;

// Visible pathElems need a vertex to be rendered
private pathElemToVertex: Map<PathElem, DdgVertex>;
// Since pathElems can share a vertex, vertices need to be tracked by their key
private vertices: Map<string, DdgVertex>;
```

I'm working on a separate branch three degrees removed so I didn't just make these changes and push, but I didn't want to lose these notes before our meeting tomorrow